mp_api/client/mprester.py

```diff
@@ -1477,3 +1477,148 @@ def query(*args, **kwargs):
+            if isinstance(entry, dict):
+                entry["uncorrected_energy_per_atom"] = entry["energy"] / sum(
+                    entry["composition"].values()
+                )
+            else:
+                entry = {
+                    "data": entry.data,
+                    "uncorrected_energy_per_atom": entry.uncorrected_energy_per_atom,
+                    "composition": entry.composition,
+                }
+
+            mp_id = entry["data"]["material_id"]
+            if (run_type := entry["data"]["run_type"]) not in energies[mp_id]:
+                energies[mp_id][run_type] = {
+                    "total_energy_per_atom": float("inf"),
+                    "composition": None,
+                }
+
+            if (
+                entry["uncorrected_energy_per_atom"]
+                < energies[mp_id][run_type]["total_energy_per_atom"]
+            ):
+                energies[mp_id][run_type] = {
+                    "total_energy_per_atom": entry["uncorrected_energy_per_atom"],
+                    "composition": entry["composition"],
+                }
+
+        atomic_energies = self.get_atom_reference_data()
+
+        e_coh_per_atom = {}
+        for mp_id, entries in energies.items():
+            if len(entries) == 0:
+                continue
+            prefered_func = sorted(list(entries), key=lambda k: entry_preference[k])[-1]
+            e_coh_per_atom[str(mp_id)] = self._get_cohesive_energy(
+                entries[prefered_func]["composition"],
+                entries[prefered_func]["total_energy_per_atom"],
+                atomic_energies[run_type_to_dfa.get(prefered_func, prefered_func)],
+                normalization=normalization,
+            )
+
+        if len(material_id) == 1:
+            return e_coh_per_atom[material_id[0]]
+        return e_coh_per_atom
+
+    @lru_cache
+    def get_atom_reference_data(
+        self,
+        funcs: tuple[str] = (
+            "PBE",
+            "SCAN",
+            "r2SCAN",
+        ),
+    ) -> dict[str, dict[str, float]]:
+        """Retrieve energies of isolated neutral atoms from MPContribs.
+
+        Args:
+            funcs ([str] or None) : list of functionals to retrieve data for.
+            Defaults to all available functionals ("PBE", "SCAN", "r2SCAN")
+            when set to None.
+
+        Returns:
+            (dict[str, dict[str, float]]) : dict containing isolated atom energies,
+            indexed first by the functionals in funcs, and second by the atom.
+        """
+        conv_fac = {"eV": 1.0, "meV": 1e-3, "µeV": 1e-6}
+
+        _atomic_energies = self.contribs.query_contributions(
+            query={"project": "isolated_atom_energies"},
+            fields=["formula", *[f"data.{dfa}.energy" for dfa in funcs]],
+        ).get("data")
+
+        atomic_energies = {dfa: {} for dfa in funcs}
+        for entry in _atomic_energies:
+            for dfa in atomic_energies:
```

Comment:
may be better to do `for dfa in funcs` here since `atomic_energies` is accessed and potentially changed as it is looped.